chalice/utils.py

```diff
@@ -96,6 +96,29 @@ def serialize_to_json(data):
     return json.dumps(data, indent=2, separators=(',', ': ')) + '\n'
 
 
+class ChaliceZipFile(zipfile.ZipFile):
+    """Support deterministic zipfile generation.
+
+    Normalizes datetime and permissions.
+    """
+
+    compression = 0  # Try to make mypy happy.
+
+    # pylint: disable=W0221
+    def write(self, filename, arcname=None, compress_type=None):
+        # type: (Text, Optional[Text], Optional[int]) -> None
+        # Only supports files, py2.7 and 3 have different signatures
+
+        info = zipfile.ZipInfo(arcname or filename)
+        # Grant other users permissions to read,
+        # else lambda blows up with mysterious error.
+        # http://unix.stackexchange.com/questions/14705/
+        info.external_attr = 0o644 << 16
```

Comment:
Should we always be setting the permissions to a specific value? Is it possible say a user wants a file in the `vendor` directory to be executable?

Reply:
So I looked into this, and yes it's possible that something in vendor is executable, and we should be preserving this.

I think the issue mentioned in the comments should be tackled separately from the issue of deterministic zip file generation.  Specifically if the file isn't readable by lambda, you'll run into issues when you run your lambda function (for example: https://github.com/aws/chalice/issues/1350#issuecomment-585453733).

All we really care about here is that we don't set the `date_time` to the mtime of the file.  So rather than duplicate the logic in `from_file` (https://github.com/python/cpython/blob/master/Lib/zipfile.py#L490) I'd rather just call `from_file` but then set the `date_time` to the default value of `(1980, 1, 1, 0, 0, 0, )` on the `ZipInfo`.  I'll also add a test that checks the file permissions.

As a separate PR, we can add logic that ensures the file is always readable.

Reply:
my two cents, we're packaging up the files here, why we wouldn't work to make the packaged files usable here seems odd, there's already tests here around permission preservation.

Reply:
I started down this path, and turns out it's much more complex than this change (especially given the parity differences between zipfile in py2 and py3) and I'd rather have smaller changes going through than block this on getting those details right.  I agree we should fix this, but I tend to prefer smaller more focused PRs.